### PR TITLE
[DateSelector] Take timezone into account

### PR DESCRIPTION
**Pull request by sebastianekstrom**
### WHY are these changes introduced?
This PR will fix an issue with the `<DateSelector />` component where it display the incorrect date in the `<TextField />` due to it not taking into account the timezone offset.

The issue is that when passing a date without time, e.g `new Date("2018-12-04").toISOString()` it doesn't take the time zone offset into account.

### WHAT is this pull request doing?
This PR fixes the issue by taking the timezone into account when adding a date either in the `<TextField />` component, or by clicking in the `<DatePicker />`. This PR also fixes https://github.com/Shopify/polaris-react/issues/688.

**Before (computer located in Sweden)**
tz-before

**After (computer located in Sweden)**
tz-after

## 

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris-react/blob/master/README.md#development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Versioning%20and%20changelog.md)

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import * as React from 'react';
import {Page} from '@shopify/polaris';
import {noop} from '@shopify/javascript-utilities/other';
import DateSelector from '../src/components/ResourceList/components/FilterControl/components/DateSelector';
interface State {}

export default class Playground extends React.Component<never, State> {
  render() {
    return (
      <Page title="Playground">
        <DateSelector
          onFilterValueChange={noop}
          onFilterKeyChange={noop}
          filterValue="on_or_before"
          filterKey="starts"
          filterMinKey="starts_min"
          filterMaxKey="starts_max"
        />
      </Page>
    );
  }
}
```

</details>

I've tested this in multiple browsers, as well as multiple time zones (setting them manually in my Mac's system preferences) and everything seems to work as intended. I'd suggest changing the time zone and clicking through the date picker and adding a date manually in the textbox.

### 🎩 checklist

* [x] Tested on [mobile](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md#cross-browser-testing)
* [x] Tested on [multiple browsers](https://help.shopify.com/en/manual/intro-to-shopify/shopify-admin/supported-browsers)
* [x] Tested for [accessibility](https://github.com/Shopify/polaris-react/blob/master/documentation/Accessibility%20testing.md)

### CODE REVIEW

File: src/components/ResourceList/components/FilterControl/components/DateSelector/DateSelector.tsx [hunk 1]
```diff
@@ -341,7 +341,8 @@ class DateSelector extends React.PureComponent<CombinedProps, State> {
       return;
     }
 
-    const nextDate = new Date(userInputDate.replace(/-/g, '/'));
+    const formattedDate = formatDateValue(new Date(userInputDate));
```
Comment (loic-d) [inline src/components/ResourceList/components/FilterControl/components/DateSelector/DateSelector.tsx]: Nitpick but I feel that `formattedDate` does not reflect completely the intention here. Same for `formatDateValue`.

Maybe the function should be called `formatDateForTimezone`?
And your const `formattedDateForTimezone`?

Comment (sebastianekstrom) [reply src/components/ResourceList/components/FilterControl/components/DateSelector/DateSelector.tsx]: That makes sense, I'll fix it

File: src/components/ResourceList/components/FilterControl/components/DateSelector/DateSelector.tsx [hunk 2]
```diff
@@ -408,8 +411,12 @@ function getDateFilterOption(
   return filterValue;
 }
 
-function formatDateValue(date: Date) {
-  return date.toISOString().slice(0, 10);
+function formatDateForTimezone(date: Date) {
+  const tzOffsetTime = new Date(
+    date.getTime() - date.getTimezoneOffset() * 60000,
```
Comment (danrosenthal) [inline src/components/ResourceList/components/FilterControl/components/DateSelector/DateSelector.tsx]: Would also be nice to move `60000` into a constant to lend some clarity as to what exactly it is.

**Comment by sebastianekstrom** [comment]
After some more manual testing I noticed some issues which boils down to that when calling `new Date("2018-08-20")` (which we input in the `<TextField />`) it actually does the following:

```
new Date("2018-08-20") -/+ Your timezones offset in hours
```

For the current code this works well in some `UTC-` countries, but not for some `UTC+X` countries. I'll dig a bit more and see if I can properly set all dates to a `2018-08-20T00:00:00` format which will make everything more robust. If you have another idea, please let me know!

**Comment by sebastianekstrom** [comment]
@danrosenthal I updated the formatting logic now, and added tests. Could you give it another look?

**Comment by danrosenthal** [comment]
Code coverage and changes look great. Can you show some evidence of this having been tophatted in `Shopify/web`?

File: UNRELEASED.md [hunk 3]
```diff
@@ -15,6 +15,7 @@ Use [the changelog guidelines](https://git.io/polaris-changelog-guidelines) to f
 - Fixed Popover not opening in a small Scrollable container ([#658](https://github.com/Shopify/polaris-react/pull/658))
 - Fixed `Page` header component to only render actions wrapper when actions are present ([#732](https://github.com/Shopify/polaris-react/pull/732))
 - Fixed `ContextualSaveBarProps` type not being exported ([#734](https://github.com/Shopify/polaris-react/pull/734))
+- Fixed timezone issue in `DateSelector` ([#710]
```
Comment (tmlayton) [inline UNRELEASED.md]:
```suggestion
- Fixed timezone issue in `DateSelector` ([#710](https://github.com/Shopify/polaris-react/pull/710))
```

File: src/components/ResourceList/components/FilterControl/components/DateSelector/DateSelector.tsx [hunk 4]
```diff
@@ -408,8 +413,22 @@ function getDateFilterOption(
...
+  if (hourOffset === 0) {
+    return date.toISOString();
+  }
+
+  const midnight =
+    hourOffset > -1
+      ? date.getHours() + hourOffset
+      : date.getHours() - hourOffset;
+  const updatedDate = new Date(date.setHours(midnight));
+  const month = padStart(String(updatedDate.getUTCMonth() + 1), 2, '0');
```
Comment (tmlayton) [inline src/components/ResourceList/components/FilterControl/components/DateSelector/DateSelector.tsx]:
```suggestion
  const month = `0${String(updatedDate.getUTCMonth() + 1)}`.slice(-2);
```

I think `padStart` is going to do [more work than we need it to](https://github.com/lodash/lodash/blob/4.17.10/lodash.js#L14442).

**Comment by dpersing** [comment]
Hi @sebastianekstrom! Can you take a look at @danrosenthal and @tmlayton's questions about test cases and tophatting in `Shopify/web`? Thank you!

**Comment by sebastianekstrom** [comment]
Tests has been updated and here's some 🎩'ing in web (discounts view)!

**Before**



before
**After**



after

ping @danrosenthal and @tmlayton

**Comment by sebastianekstrom** [comment]
Noticed a few edge cases when 🎩ing in `web`, not really sure what causes it but I'm gonna look into the issue a bit further!

**Comment by sebastianekstrom** [comment]
Web 🎩 ing:

**Timezone offset: -720**

**Timezone offset: -480**

**Timezone offset: -60**

**Timezone offset: 0**

**Timezone offset: 300**


Everything seems to work great!
Running tests still seems to be dependent on the local timezone even though I'm mocking it. Investigating how I can improve the mocking.

File: src/components/ResourceList/components/FilterControl/components/DateSelector/DateSelector.tsx [hunk 5]
```diff
@@ -408,8 +412,21 @@ function getDateFilterOption(
   return filterValue;
 }
 
-function formatDateValue(date: Date) {
-  return date.toISOString().slice(0, 10);
+function formatDateForTimezone(date: Date) {
```
Comment (loic-d) [inline src/components/ResourceList/components/FilterControl/components/DateSelector/DateSelector.tsx]: Can you use one of the date helpers in Quilt?
https://github.com/Shopify/quilt/tree/master/packages/dates

There is an helper to apply a timezone offset on a date. Maybe it could work?

File: src/components/ResourceList/components/FilterControl/components/DateSelector/tests/DateSelector.test.tsx [hunk 6]
```diff
@@ -161,6 +161,37 @@ describe('<DateSelector />', () => {
...
 
+  describe.only('timezones adjustments', () => {
+    const timeZoneOffsets = [720, 540, 60, 0, -60, -540, -720];
+    const dates = ['2017-12-01', '2019-01-01', '2019-08-22'];
+
+    const getTimezoneOffset = Date.prototype.getTimezoneOffset;
+
+    dates.forEach((date) => {
+      timeZoneOffsets.forEach((zone) => {
+        it(`sets the correct values for ${date} with timezone offset ${zone}`, () => {
```
Comment (loic-d) [inline src/components/ResourceList/components/FilterControl/components/DateSelector/tests/DateSelector.test.tsx]: It feels a bit weird to have a test in nested `forEach`.

I would instead run one test for a given date and given offset. If it works for one offset, it should work for all offsets no?

Your `getTimezoneOffset` mock does not seem to be properly set.

If you want a mock for multiple tests you could do something like this I think?

```
beforeEach(() => {
    getTimezoneOffset = jest.spyOn(Date.prototype, 'getTimezoneOffset');
  });

  afterEach(() => {
    getTimezoneOffset.mockReset();
  });

In your test: 

getTimezoneOffset.mockImplementation(() => mockedOffset);

```

File: src/components/ResourceList/components/FilterControl/components/DateSelector/DateSelector.tsx [hunk 7]
```diff
@@ -408,8 +411,22 @@ function getDateFilterOption(
...
+  const timezoneOffset = date.getTimezoneOffset() * 60 * 1000;
```
Comment (loic-d) [inline src/components/ResourceList/components/FilterControl/components/DateSelector/DateSelector.tsx]: Maybe `timezoneOffsetMs`?

File: src/components/ResourceList/components/FilterControl/components/DateSelector/DateSelector.tsx [hunk 8]
```diff
@@ -408,8 +411,22 @@ function getDateFilterOption(
...
+  if (date.getTimezoneOffset() > 0) {
+    const newTime = date.getTime() + timezoneOffset;
+    date.setTime(newTime);
```
Comment (loic-d) [inline src/components/ResourceList/components/FilterControl/components/DateSelector/DateSelector.tsx]: I think your function should not mutate the received date, but instead return a new `Date` object?

File: src/components/ResourceList/components/FilterControl/components/DateSelector/DateSelector.tsx [hunk 9]
```diff
@@ -408,8 +411,22 @@ function getDateFilterOption(
...
+  } else {
+    const newTime = date.getTime() - timezoneOffset;
+    date.setTime(newTime);
+  }
+  return date.toISOString();
+}
+
+function formatDateForInput(date: Date) {
+  const month = `0${String(date.getMonth() + 1)}`.slice(-2);
+  const day = `0${String(date.getDate())}`.slice(-2);
```
Comment (loic-d) [inline src/components/ResourceList/components/FilterControl/components/DateSelector/DateSelector.tsx]: no timezone adjustment needed here?

File: src/components/ResourceList/components/FilterControl/components/DateSelector/tests/DateSelector.test.tsx [hunk 10]
```diff
@@ -161,6 +161,32 @@ describe('<DateSelector />', () => {
     });
   });
 
+  describe('timezones adjustments', () => {
+    const timeZoneOffsets = [720, 540, 60, 0, -60, -540, -720];
+    const date = '2019-01-01';
+    const getTimezoneOffset = Date.prototype.getTimezoneOffset;
+
+    timeZoneOffsets.forEach((zone) => {
```
Comment (loic-d) [inline src/components/ResourceList/components/FilterControl/components/DateSelector/tests/DateSelector.test.tsx]: I would also 🔥 this `forEach` here, if it works for one offset it should work for all offsets no?
Assuming your mock is properly called

File: src/components/ResourceList/components/FilterControl/components/DateSelector/tests/DateSelector.test.tsx [hunk 11]
```diff
@@ -161,6 +161,32 @@ describe('<DateSelector />', () => {
...
+      it(`sets the correct values for ${date} with timezone offset ${zone}`, () => {
+        Date.prototype.getTimezoneOffset = () => zone; // eslint-disable-line no-extend-native
```
Comment (loic-d) [inline src/components/ResourceList/components/FilterControl/components/DateSelector/tests/DateSelector.test.tsx]: Is this allowed? Can't you use `mockImplementation`? it feels a bit weird to overwrite it like that.

File: src/components/ResourceList/components/FilterControl/components/DateSelector/DateSelector.tsx [hunk 12]
```diff
@@ -273,15 +273,15 @@ class DateSelector extends React.PureComponent<CombinedProps, State> {
...
+        selectedDate ? formatDateForTimezone(selectedDate) : undefined,
       );
       return;
     }
 
     if (newOption === DateFilterOption.OnOrAfter) {
       onFilterKeyChange(filterMinKey);
       onFilterValueChange(
-        selectedDate ? formatDateValue(selectedDate) : undefined,
+        selectedDate ? formatDateForTimezone(selectedDate) : undefined,
```
Comment (loic-d) [inline src/components/ResourceList/components/FilterControl/components/DateSelector/DateSelector.tsx]: Maybe you could store the result of `formatDateForTimezone` into a constant since you are using it twice?

File: src/components/ResourceList/components/FilterControl/components/DateSelector/DateSelector.tsx [hunk 13]
```diff
@@ -341,7 +341,11 @@ class DateSelector extends React.PureComponent<CombinedProps, State> {
       return;
     }
 
-    const nextDate = new Date(userInputDate.replace(/-/g, '/'));
+    const formattedDateForTimezone = formatDateForTimezone(
+      new Date(userInputDate),
+    );
+
+    const nextDate = new Date(formattedDateForTimezone);
```
Comment (loic-d) [inline src/components/ResourceList/components/FilterControl/components/DateSelector/DateSelector.tsx]: `formatDateForTimezone` could return a new date object directly.

File: src/components/ResourceList/components/FilterControl/components/DateSelector/DateSelector.tsx (hunk 5 above)
Comment (loic-d) [reply src/components/ResourceList/components/FilterControl/components/DateSelector/DateSelector.tsx]: As discussed the date package is not available in Polaris.

File: src/components/ResourceList/components/FilterControl/components/DateSelector/DateSelector.tsx [hunk 14]
```diff
@@ -408,8 +411,19 @@ function getDateFilterOption(
   return filterValue;
 }
 
-function formatDateValue(date: Date) {
-  return date.toISOString().slice(0, 10);
+function formatDateForTimezone(date: Date) {
```
Comment (loic-d) [inline src/components/ResourceList/components/FilterControl/components/DateSelector/DateSelector.tsx]: I feel this function name could be improved a little?
It's not clear which timezone it's using unless you look at the implementation.

File: src/components/ResourceList/components/FilterControl/components/DateSelector/DateSelector.tsx [hunk 15]
```diff
@@ -408,8 +411,19 @@ function getDateFilterOption(
...
+  const timezoneOffsetMs = date.getTimezoneOffset() * 60 * 1000;
+  const formattedDate = new Date();
+
+  if (date.getTimezoneOffset() > 0) {
+    const newTime = date.getTime() + timezoneOffsetMs;
+    formattedDate.setTime(newTime);
```
Comment (loic-d) [inline src/components/ResourceList/components/FilterControl/components/DateSelector/DateSelector.tsx]: Personally I prefer early returns over `if/else`

File: src/components/ResourceList/components/FilterControl/components/DateSelector/tests/DateSelector.test.tsx [hunk 16]
```diff
@@ -161,6 +161,35 @@ describe('<DateSelector />', () => {
     });
   });
 
+  describe('timezones adjustments', () => {
+    const date = '2019-01-01';
+    let getTimezoneOffset: jest.SpyInstance;
+
+    beforeEach(() => {
+      getTimezoneOffset = jest.spyOn(Date.prototype, 'getTimezoneOffset');
```
Comment (loic-d) [inline src/components/ResourceList/components/FilterControl/components/DateSelector/tests/DateSelector.test.tsx]: Nice! If you are only mocking it for one test, it can be done in the test itself.

File: src/components/ResourceList/components/FilterControl/components/DateSelector/tests/DateSelector.test.tsx [hunk 17]
```diff
@@ -161,6 +161,35 @@ describe('<DateSelector />', () => {
...
+    });
+
+    afterEach(() => {
+      getTimezoneOffset.mockRestore();
+    });
+
+    it(`sets the correct values for ${date} with timezone offset -540`, () => {
```
Comment (loic-d) [inline src/components/ResourceList/components/FilterControl/components/DateSelector/tests/DateSelector.test.tsx]: No need to interpolate here. 
It could just be `sets the correct selected date on DatePicker for date with timezone offset`?

File: src/components/ResourceList/components/FilterControl/components/DateSelector/tests/DateSelector.test.tsx [hunk 18]
```diff
@@ -161,6 +161,35 @@ describe('<DateSelector />', () => {
...
+          {...mockDefaultProps}
+          filterValue={DateFilterOption.OnOrBefore}
+        />,
+      );
+
+      trigger(wrapper.find(TextField), 'onChange', date);
+      trigger(wrapper.find(TextField), 'onBlur');
+
+      const selectedDate = wrapper.find(DatePicker).prop('selected') as Date;
+      expect(selectedDate.toISOString()).toContain(date);
```
Comment (loic-d) [inline src/components/ResourceList/components/FilterControl/components/DateSelector/tests/DateSelector.test.tsx]: Shouldn't your offset be used in the assertion?

File: src/components/ResourceList/components/FilterControl/components/DateSelector/tests/DateSelector.test.tsx [hunk 19]
```diff
@@ -179,6 +208,15 @@ describe('<DateSelector />', () => {
   });
 
   describe('onFilterValueChange', () => {
+    let getTimezoneOffset: jest.SpyInstance;
+    beforeEach(() => {
+      getTimezoneOffset = jest.spyOn(Date.prototype, 'getTimezoneOffset');
+    });
```
Comment (loic-d) [inline src/components/ResourceList/components/FilterControl/components/DateSelector/tests/DateSelector.test.tsx]: You could move the mock one level up, this way you don't have to define it twice for two different `describe`

File: src/components/ResourceList/components/FilterControl/components/DateSelector/tests/DateSelector.test.tsx [hunk 20]
```diff
@@ -231,6 +269,7 @@ describe('<DateSelector />', () => {
       const onFilterValueChangeSpy = jest.fn();
       const newDateFilter = DateFilterOption.OnOrAfter;
       const date = '2019-05-28';
+      getTimezoneOffset.mockImplementation(() => -540);
```
Comment (loic-d) [inline src/components/ResourceList/components/FilterControl/components/DateSelector/tests/DateSelector.test.tsx]: You should use the offset in the assertion somehow.

File: src/components/ResourceList/components/FilterControl/components/DateSelector/tests/DateSelector.test.tsx [hunk 21]
```diff
@@ -299,7 +347,9 @@ describe('<DateSelector />', () => {
       trigger(wrapper.find(TextField), 'onChange', date);
       trigger(wrapper.find(TextField), 'onBlur');
 
-      expect(onFilterValueChangeSpy).toHaveBeenCalledWith(date);
+      expect(onFilterValueChangeSpy).toHaveBeenCalledWith(
+        '2019-08-22T18:00:00.000Z',
```
Comment (loic-d) [inline src/components/ResourceList/components/FilterControl/components/DateSelector/tests/DateSelector.test.tsx]: I feel that the expected values should be built using your mocked timezone offset

**Comment by sebastianekstrom** [comment]
@loic-d all comments should be fixed now!

File: src/components/ResourceList/components/FilterControl/components/DateSelector/tests/DateSelector.test.tsx [hunk 22]
```diff
@@ -161,6 +172,48 @@ describe('<DateSelector />', () => {
     });
   });
 
+  describe('timezones adjustments', () => {
+    const date = '2019-01-01';
+
+    it('sets the correct selected date with timezone offset -540 on DatePicker and TextField', () => {
+      getTimezoneOffset.mockImplementation(() => -540);
+      const wrapper = mountWithAppProvider(
```
Comment (loic-d) [inline src/components/ResourceList/components/FilterControl/components/DateSelector/tests/DateSelector.test.tsx]: Better to have a line break for readability:

- Test setup
- Mount component
- Assert

File: src/components/ResourceList/components/FilterControl/components/DateSelector/tests/DateSelector.test.tsx [hunk 23]
```diff
@@ -161,6 +172,48 @@ describe('<DateSelector />', () => {
...
+        />,
+      );
+
+      trigger(wrapper.find(TextField), 'onChange', date);
+      trigger(wrapper.find(TextField), 'onBlur');
+
+      const selectedDate = wrapper.find(DatePicker).prop('selected') as Date;
+      const selectedInputDate = wrapper.find(TextField).prop('value');
+
+      expect(selectedDate.toISOString()).toBe('2019-01-01T09:00:00.000Z');
```
Comment (loic-d) [inline src/components/ResourceList/components/FilterControl/components/DateSelector/tests/DateSelector.test.tsx]: Does the timezone offset you mocked have an impact on `2019-01-01T09:00:00.000Z`? If yes, you should build this date string using the mocked offset.

File: src/components/ResourceList/components/FilterControl/components/DateSelector/tests/DateSelector.test.tsx [hunk 24]
```diff
@@ -161,6 +172,48 @@ describe('<DateSelector />', () => {
     });
   });
 
+  describe('timezones adjustments', () => {
+    const date = '2019-01-01';
+
+    it('sets the correct selected date with timezone offset -540 on DatePicker and TextField', () => {
```
Comment (loic-d) [inline src/components/ResourceList/components/FilterControl/components/DateSelector/tests/DateSelector.test.tsx]: ... with negative timezone offset ...

File: src/components/ResourceList/components/FilterControl/components/DateSelector/tests/DateSelector.test.tsx [hunk 25]
```diff
@@ -161,6 +172,48 @@ describe('<DateSelector />', () => {
...
+      trigger(wrapper.find(TextField), 'onBlur');
+
+      const selectedDate = wrapper.find(DatePicker).prop('selected') as Date;
+      const selectedInputDate = wrapper.find(TextField).prop('value');
+
+      expect(selectedDate.toISOString()).toBe('2019-01-01T09:00:00.000Z');
+      expect(selectedInputDate).toBe(date);
+    });
+
+    it('sets the correct selected date with timezone offset 300 on DatePicker and TextField', () => {
```
Comment (loic-d) [inline src/components/ResourceList/components/FilterControl/components/DateSelector/tests/DateSelector.test.tsx]: ... with positive timezone offset ...

Otherwise it sounds like it will only work for the specific offsets you mocked.

File: src/components/ResourceList/components/FilterControl/components/DateSelector/tests/DateSelector.test.tsx [hunk 26]
```diff
@@ -161,6 +172,48 @@ describe('<DateSelector />', () => {
...
+        />,
+      );
+
+      trigger(wrapper.find(TextField), 'onChange', date);
+      trigger(wrapper.find(TextField), 'onBlur');
+
+      const selectedDate = wrapper.find(DatePicker).prop('selected') as Date;
+      const selectedInputDate = wrapper.find(TextField).prop('value');
+
+      expect(selectedDate.toISOString()).toBe('2019-01-01T05:00:00.000Z');
```
Comment (loic-d) [inline src/components/ResourceList/components/FilterControl/components/DateSelector/tests/DateSelector.test.tsx]: Same here, does the timezone offset you mocked have an impact of the selected date?

**Comment by loic-d** [comment]
Thank you for integrating the feedback!

Looks good. I added some more other minor comments. In the tests, I still feel that there is something to do with the mocked timezone offset. You are mocking it but are not using the offset in the assertions.

I'll leave it to @danrosenthal and @tmlayton for the final ✅

File: src/components/ResourceList/components/FilterControl/components/DateSelector/DateSelector.tsx [hunk 27]
```diff
@@ -408,8 +413,19 @@ function getDateFilterOption(
...
+    return formattedDate.toISOString();
+  }
+
+  const newTime = date.getTime() - timezoneOffsetMs;
```
[the remark this answers is not in the document]
Comment (sebastianekstrom) [reply src/components/ResourceList/components/FilterControl/components/DateSelector/DateSelector.tsx]: I don't think that's possible unfortunately, since when it's a negative timezone offset we want to _subtract_ and when it's a positive we want to _add_.
